### PR TITLE
feat(eslint-plugin): add return-await rule

### LINKED ISSUE #994: Feature request - rule for await in async try catch return
A `Rejected Promise` returned by an `async` function will not be caught by the function's `try-catch` block. This may surprise some users and lead to unexpected behavior:
```typescript
(async () => {
  try {
    // will result in an UnhandledPromiseRejection
    return Promise.reject('Error!');
  } catch (e) {
    // will not run
    handleCleanup();
  }
})();
```

conversely, `await`-ed promises work in a more intuitive fashion:
```typescript
(async () => {
  try {
    // will not result in an UnhandledPromiseRejection
    return await Promise.reject('Error!');
  } catch (e) {
    // will run
    handleCleanup();
  }
})();
```

I understand why this is the case according to the `async-await` function spec, but it feels like confusing behavior and a problem that `typescript-eslint` could help avoid via the following rule:

disallow `return`-ing a `Promise` from inside an `async` function `try-catch` block.

Thus, the first case described above would be invalid according to this rule, but the second case would be valid. We could even, potentially, provide an auto-fix that would add the `await` statement to the promise.

There's also the possibility for the broader rule of disallowing`return`-ing a `Promise` from anywhere inside an `async` function, but I wanted to start with the narrowest case.

Please let me know if there's any other clarifying information I can provide, thanks so much for your time!

### PR DESCRIPTION
**Pull request by drabinowitz**
This closes #994 adding the `return-await` rule.

### CODE REVIEW

**Comment by drabinowitz** [comment]
@bradzacher just wanted to check in on this to see if any other info was needed from my end.

**Comment by bradzacher** [comment]
I haven't had a look, sorry - it's in the review queue, so I'll take a look soon.

**Comment by drabinowitz** [comment]
Sounds good no rush, just wanted to make sure there wasn't anything missing I needed to take care of before it was eligible for review.

**Comment by drabinowitz** [comment]
@bradzacher i wanted to check in on this pr to see if you'd had a chance to take a look at it.

**Comment by bradzacher** [comment]
I have not yet. My open source time lately has been consumed by fixing the IDE errors.
It's on the list of PRs to review.

File: packages/eslint-plugin/src/rules/return-await.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import ts, { SyntaxKind } from 'typescript';
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'return-await',
+  meta: {
+    docs: {
+      description: 'Rules for awaiting returned promises',
+      category: 'Best Practices',
+      recommended: 'error',
```
[the remark this answers is not in the document]
Comment (drabinowitz) [reply packages/eslint-plugin/src/rules/return-await.ts]: will do!

**Pull request merged**